File: frontend/types/dns-record.ts
export type RecordType = "A" | "AAAA" | "CNAME" | "TXT" | "MX" | "NS" | "PTR" | "SRV" | "CAA";
export type RoutingPolicy = "SIMPLE" | "WEIGHTED";

export interface DnsRecord {
  id: string;
  hosted_zone_id: string;
  name: string;
  type: RecordType;
  ttl: number;
  values: (string | Record<string, unknown>)[];
  routing_policy: RoutingPolicy;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface DnsRecordCreate {
  name: string;
  type: RecordType;
  ttl: number;
  values: (string | Record<string, unknown>)[];
  routing_policy?: RoutingPolicy;
}

export interface DnsRecordUpdate {
  ttl?: number;
  values?: (string | Record<string, unknown>)[];
  routing_policy?: RoutingPolicy;
}

export const RECORD_TYPES: RecordType[] = ["A", "AAAA", "CNAME", "TXT", "MX", "NS", "PTR", "SRV", "CAA"];

File: backend/app/schemas/dns_record.py
from pydantic import BaseModel, field_validator
from typing import Optional, List, Any
from datetime import datetime


class DnsRecordCreate(BaseModel):
    name: str
    type: str
    ttl: int = 300
    values: List[Any]
    routing_policy: str = "SIMPLE"

    @field_validator("ttl")
    @classmethod
    def ttl_range(cls, v: int) -> int:
        if not (0 <= v <= 172800):
            raise ValueError("TTL must be between 0 and 172800 seconds")
        return v

    @field_validator("type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        allowed = ["A", "AAAA", "CNAME", "TXT", "MX", "NS", "PTR", "SRV", "CAA"]
        if v.upper() not in allowed:
            raise ValueError(f"Type must be one of {allowed}")
        return v.upper()

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        return v.strip()


class DnsRecordUpdate(BaseModel):
    ttl: Optional[int] = None
    values: Optional[List[Any]] = None
    routing_policy: Optional[str] = None

    @field_validator("ttl")
    @classmethod
    def ttl_range(cls, v: Optional[int]) -> Optional[int]:
        if v is not None and not (0 <= v <= 172800):
            raise ValueError("TTL must be between 0 and 172800 seconds")
        return v


class DnsRecordResponse(BaseModel):
    id: str
    hosted_zone_id: str
    name: str
    type: str
    ttl: int
    values: List[Any]
    routing_policy: str
    is_system: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
